presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkAdaptiveQueryExecution.java

```diff
@@ -0,0 +1,351 @@
+                if (frFuture.isPresent()) {
+                    SimpleFutureAction<MapOutputStatistics> fragmentFuture = frFuture.get();
+
+                    fragmentFuture.onComplete(new AbstractFunction1<Try<MapOutputStatistics>, Void>() {
+                        @Override
+                        public Void apply(Try<MapOutputStatistics> res) {
+                            if (res.isSuccess()) {
+                                // TODO Is it possible that the mapOutStats here are null? If not, we don't need to have Optional here and in the RuntimeStatistics.createRuntimeStats().
+                                Optional<MapOutputStatistics> mapOutStats = Optional.ofNullable(res.get());
+                                try {
+                                    fragmentEventQueue.put(new FragmentCompletionSuccessEvent(fragment.getFragment().getId(), mapOutStats));
+                                }
+                                catch (InterruptedException e) {
+                                    throw new RuntimeException(e);
+                                }
+                            } else {
+                                Throwable thrw = res.failed().get();
+                                try {
+                                    fragmentEventQueue.put(new FragmentCompletionFailureEvent(fragment.getFragment().getId(), thrw));
+                                }
+                                catch (InterruptedException e) {
+                                    throw new RuntimeException(e);
+                                }
+                            }
+                            return null;
+                        }
+                    }, execService);
+
+                } else {
+                    log.info("Fragment %s will not get executed now either because there was no exchange involved (a broadcast is present) or because of another issue.",
+                            fragment.getFragment().getId());
+                }
+
+            }
+
+            // Consume the next fragment execution completion event (block if no new fragment execution has finished) and re-optimize if possible.
+            // TODO Put this in a separate thread?
+            FragmentCompletionEvent fragmentEvent;
+            try {
+                fragmentEvent = fragmentEventQueue.take();
+            }
+            catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+            if (fragmentEvent instanceof FragmentCompletionSuccessEvent) {
+                FragmentCompletionSuccessEvent successEvent = (FragmentCompletionSuccessEvent) fragmentEvent;
+                executedFragments.add(successEvent.getFragmentId());
+                Optional<PlanNodeStatsEstimate> runtimeStats = RuntimeStatistics.createRuntimeStats(successEvent.getMapOutputStats());
+                // Re-optimizations here.
+
+            } else if (fragmentEvent instanceof FragmentCompletionFailureEvent) {
```

Comment:
I think it would be clearer to change this logic around a bit, since all the rest of the method assumes a success event.  Changing it would also ensure that we don't end up with an unhandled type of fragmentCompletionEvent.
```
if(fragmentEvent instanceof FragmentCompletionFailureEvent) {
   throw ...
}

verify(fragmentEvent instance of FragmentCompletionSuccessEvent, "Unexpected FragmentCompletionEvent type: %s", fragmentEvent.class.getSimpleName());
...
```